A41-xds-rbac.md

```diff
@@ -0,0 +1,230 @@
+are less common than client-side at this point in time, so this may be the first
+usage of `RouteConfiguration` on server-side and so likely involves adding more
+complete support for A39 by creating and executing filters on server-side and
+processing `VirtualHost`s and `RoutMatch`es to determine which configuration
+should be provided to each filter. A39 should be consulted for the expected
+behavior.
+
+New validation should occur for `HttpConnectionManager` to allow equating
+RBAC's `direct_remote_ip` and `remote_ip`. If the RBAC implementation does not
+distinguish between these fields, then
+`HttpConnectionManager.xff_num_trusted_hops` must be unset or zero and
+`HttpConnectionManager.original_ip_detection_extensions` must be empty. If
+either field has an incorrect value, the Listener must be NACKed. For
+simplicity, this behavior applies independent of the Listener type (both
+client-side and server-side).
+
+The core policy matching logic should be split into an "RBAC engine" to allow
+internal reuse with a non-xDS API. Any non-xDS API will not be RBAC, so this
+reuse is merely an implementation detail. The API is not part of this gRFC.
+There is no requirement that the RBAC engine have a specialized API; it could
+simply be an interceptor.
+
+The [RBAC policy][] has many fields. All current (Envoy-implemented) fields will
+be considered in gRPC. However, some fields may have pre-determined values or
+behavior. At this time, if the `RBAC.action` is `Action.LOG` then the policy
+will be completely ignored, as if RBAC was not configurated. CEL fields are not
+supported, so `Policy.condition` and `Policy.checked_condition` must cause a
+validation failure if present. It is also a validation failure if Permission or
+Principal has a `header` matcher for a `grpc-`-prefixed header name or
+`:scheme`. As described in A39, validation failures for filter configuration
+causes the listener to be NACKed.
+
+The following fields of `Permission` may not be obvious how they map to gRPC:
+
+| Permission Field | gRPC Equivalent |
+| ---------------- | --------------- |
+| header           | Metadata (with caveats below) |
+| url_path         | Fully-qualified RPC method name with leading slash. Same as `:path` header |
+| destination_ip   | Local address for this connection |
+| destination_port | Local port for this connection |
+| metadata         | Hard-coded as empty; never matches |
+| requested_server_name | Hard-coded as empty string |
+
+Because matching supports NOT, the matcher must still be processed even if a
+rule contains references to things that don't generally match; it is not trivial
+to "optimize out" the never-matching rules.
+
+The `header` field is not entirely 1:1 with gRPC Metadata, in part because which
+HTTP headers are present in Metadata is not 100% consistent cross-language.
+For this design, `headers` can include `:method`, `:authority`, and `:path`
```

Comment:
No normalizations are done on the path. But no normalizations are done by the grpc service either. gRPC essentially just does string-equality to match RPC methods. (It is possible for applications to have some freedom here, but nobody really does differently.)

Envoy has a bigger problem here since the HTTP library processing the request in the application is a different one than in Envoy. That isn't really a problem in a proxyless environment, as long as we make sure RBAC runs with the same path that the service would see.